proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,221 @@
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />` to remove any `auditSources` inherited from parent *nuget.config* files.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+Omitting `<clear />` will not use `packageSources` in addition to `auditSources`.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <!-- Clear to ensure package sources not inherited from other config files -->
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
+  </packageSources>
+  <auditSources>
+    <!-- Clear to ensure audit sources not inherited from other config files -->
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
+  </auditSources>
+</configuration>
+```
+
+Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
+However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.
+
+NuGet will use same semantics for `auditSources` as it already has with `packageSources`.
```

Comment:
You established [here](https://github.com/NuGet/Home/pull/12918/files#diff-d4c82f8e89642e3f05039c1cb46e315bb78bf6ba400e7c6a6690e2fabc464413R61) that audit sources must implement the V3 protocol.

Package sources support a [`protocolVersion`](https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#packagesources) attribute.  You should be explicit about the behavior if an audit source has a `protocolVersion` attribute with a value other than 3 (e.g.:  2).  Do we ignore, warn, or error?

Reply:
+1 on this point. I think this is a case where `<auditSources>` has difference semantics from `<packageSources>`. `protocolVersion` must be excluded or set to `3` and file-based sources should be rejected right?

IMO an error is best for this case since it's a new config value and we can therefore be strict from the beginning.

Reply:
If we're going to re-introduce NU1905 for when an audit source doesn't provide the VulnerabilityInfo resource, then it'll happen "automatically". The warning or error message might not be specifically calling out that v2 sources don't support it, but that's something we can document on the NU1905 docs page.